cpp/src/neighbors/detail/hnsw.hpp

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuvs/neighbors/hnsw.hpp>
+#include <filesystem>
+#include <hnswlib/hnswalg.h>
+#include <hnswlib/hnswlib.h>
+#include <memory>
+#include <random>
+
+namespace cuvs::neighbors::hnsw::detail {
+
+template <typename T>
+struct hnsw_dist_t {
+  using type = void;
+};
+
+template <>
+struct hnsw_dist_t<float> {
+  using type = float;
+};
+
+template <>
+struct hnsw_dist_t<uint8_t> {
+  using type = int;
+};
+
+template <>
+struct hnsw_dist_t<int8_t> {
+  using type = int;
+};
+
+template <typename T>
+struct index_impl : index<T> {
+ public:
+  /**
+   * @brief load a base-layer-only hnswlib index originally saved from a built CAGRA index
+   *
+   * @param[in] filepath path to the index
+   * @param[in] dim dimensions of the training dataset
+   * @param[in] metric distance metric to search. Supported metrics ("L2Expanded", "InnerProduct")
+   */
+  index_impl(std::string filepath, int dim, cuvs::distance::DistanceType metric)
```

Comment:
nitpick: pass by reference
```suggestion
  index_impl(const std::string & filepath, int dim, cuvs::distance::DistanceType metric)
```

Reply:
gah - I messed the style up there , sorry about that